sklearn/cluster/optics.py

```diff
@@ -0,0 +1,638 @@
+
+    localMaximaPoints = _find_local_maxima(RPlot, RPoints, nghsize)
+
+    rootNode = TreeNode(RPoints, 0, len(RPoints), None)
+    _cluster_tree(rootNode, None, localMaximaPoints,
+                  RPlot, RPoints, min_cluster_size,
+                  maxima_ratio, rejection_ratio,
+                  similarity_threshold, significant_min)
+
+    return rootNode
+
+
+class TreeNode(object):
+    # automatic cluster helper classes and functions
+    def __init__(self, points, start, end, parentNode):
+        self.points = points
+        self.start = start
+        self.end = end
+        self.parentNode = parentNode
+        self.children = []
+        self.splitpoint = -1
+
+#    def __str__(self):
+#        return "start: %d, end %d, split: %d" % (self.start,
+#                                                 self.end,
+#                                                 self.splitpoint)
+
+    def assignSplitPoint(self, splitpoint):
+        self.splitpoint = splitpoint
+
+    def addChild(self, child):
+        self.children.append(child)
+
+
+def _is_local_maxima(index, RPlot, RPoints, nghsize):
+    # 0 = point at index is not local maxima
+    # 1 = point at index is local maxima
+
+    for i in range(1, nghsize+1):
+        # process objects to the right of index
+        if index + i < len(RPlot):
+            if (RPlot[index] < RPlot[index+i]):
+                return 0
+
+        # process objects to the left of index
+        if index - i >= 0:
+            if (RPlot[index] < RPlot[index-i]):
+                return 0
+
+    return 1
+
+
+def _find_local_maxima(RPlot, RPoints, nghsize):
+
+    localMaximaPoints = {}
+
+    # 1st and last points on Reachability Plot are not taken
+    # as local maxima points
+    for i in range(1, len(RPoints)-1):
+        # if the point is a local maxima on the reachability plot with
+        # regard to nghsize, insert it into priority queue and maxima list
+        if (RPlot[i] > RPlot[i-1] and RPlot[i] >= RPlot[i+1] and
+                _is_local_maxima(i, RPlot, RPoints, nghsize) == 1):
+            localMaximaPoints[i] = RPlot[i]
+
+    return sorted(localMaximaPoints,
+                  key=localMaximaPoints.__getitem__,
+                  reverse=True)
+
+
+def _cluster_tree(node, parentNode, localMaximaPoints,
+                  RPlot, RPoints, min_cluster_size,
+                  maxima_ratio, rejection_ratio,
+                  similarity_threshold, significant_min):
+    # node is a node or the root of the tree in the first call
+    # parentNode is parent node of N or None if node is root of the tree
+    # localMaximaPoints is list of local maxima points sorted in
+    # descending order of reachability
+    if len(localMaximaPoints) == 0:
+        return  # parentNode is a leaf
+
+    # take largest local maximum as possible separation between clusters
+    s = localMaximaPoints[0]
+    node.assignSplitPoint(s)
+    localMaximaPoints = localMaximaPoints[1:]
```

Comment:
Pl. convert all `camelCased` variable names to `underscore_separated_all_lower_cased_names`...